[RFC] Soundwire machine driver autodetect

plbossart opened the pull request:
This is a tentative proposal to show how we can automagically detect what SoundWire-based configuration to use for machine drivers and topology files.

The proposal is to filter a new set of tables based on a 'link_mask' key. This mask describes which links are required for a given configuration to work, which gives us a very simple first-order solution.
today we have 
CML: link1 is required for RT700
ICL: link0 is required for RT700, all 4 links are required for RT711-RT1308(x2)-RT715.

@bardliao @RanderWang please check if this works for you? @bardliao you'd need to add a dedicated machine driver for the 3-in-1, instead of trying to patch cnl_rt700.

This is not likely to scale and the link_mask key would likely need to be augmented by a list of _ADR fields expected on each link if (or rather when) when people start swapping things out.

Comments welcome, this was just an idea I wanted to check and it seems to be good enough for the initial SoundWire releases.

RanderWang commented on sound/soc/intel/common/soc-acpi-intel-icl-match.c:
```diff
@@ -29,15 +29,26 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
+		.link_mask = 0xF, /* 4 active links required */
+		.drv_name = "cnl_rt711_rt1308_rt715",
+		.sof_fw_filename = "sof-icl.ri",
+		.sof_tplg_filename = "sof-icl-rt711_rt1308_rt715.tplg",
```
CML will support 4 links, so please also add this item to CNL

plbossart replied:
done.

RanderWang commented:
I prefer most part of this PR, only one concern: Do we need to add another two machine tables and alt_machines ?  I think the new sdw machine table is still based on  ACPI information, so it should be included in original acpi table.

bardliao commented on sound/soc/sof/intel/hda.c:
```diff
@@ -517,6 +521,41 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
+			if (mach->link_mask == link_mask)
+				break;
+			mach++;
+		}
```
``While (mach)`` doesn't work. The latest mach of pdata->desc->alt_machines is not null. 
My suggestion will be
```
		for (mach = pdata->desc->alt_machines; mach->link_mask; mach++) {
			if (mach->link_mask == link_mask) {
				dev_dbg(bus->dev,
					"SoundWire machine driver %s topology %\s\n",
					mach->drv_name,
					mach->sof_tplg_filename);
				pdata->machine = mach;
				mach->mach_params.platform = dev_name(sdev->dev);
				pdata->fw_filename = mach->sof_fw_filename;
				pdata->tplg_filename = mach->sof_tplg_filename;
				break;
			}
		}
		if (!mach->link_mask)
			mach = NULL;
```

plbossart replied:
yes good catch @bardliao, even in the I2S case we do this:
````c
	for (mach = machines; mach->id[0]; mach++) {
````

Will refine this.

plbossart replied:
Maybe more compact even:

````c
if (!pdata->machine && link_mask != 0) {
		/* SoundWire enabled, revisit machine driver selection */
		mach = pdata->desc->alt_machines;
		while (mach->link_mask && mach->link_mask != link_mask)
			mach++;
		}
                if (!mach->link_mask)
			mach = NULL;
}
````

plbossart commented:
> I prefer most part of this PR, only one concern: Do we need to add another two machine tables and alt_machines ? I think the new sdw machine table is still based on ACPI information, so it should be included in original acpi table.

I had the same hesitation, but the problem is that we don't want to parse the SoundWire-related parts when looking for I2S codecs, and conversely when checking for SoundWire-based machines we don't want to look for HIDs that are related to I2S codecs.
If we merge the two tables we'd need to add a field that indicates what type each entry is, so that's quite invasive since we'd need to modify the existing code and introduce a new flag to discard entries.
Adding two types of tables is less invasive since it's only introduced for CNL+, so we leave the legacy alone.

plbossart commented:
> @plbossart typo in the commit message "enouhg"

fixed, thanks for spotting this.

plbossart commented:
updated branch, @bardliao can you try on your ICL w/ 3-in-1? I added a dedicated machine driver so in theory it would be loaded. I have some lab issues so could test if this works at all, only the rt700 case was fine.
